internal/oci/container.go

```diff
@@ -761,3 +763,49 @@ func (c *Container) RuntimePathForPlatform(r *runtimeOCI) string {
 	}
 	return c.runtimePath
 }
+
+// AddExecPID registers a PID associated with an exec session.
+// It is tracked so exec sessions can be cancelled when the container is being stopped.
+// If the PID is conmon, shouldKill should be false, as we should not call SIGKILL on conmon
+// If it is an exec session, shouldKill should be true, as we can't guarantee the exec process
+// will have a SIGINT handler.
+func (c *Container) AddExecPID(pid int, shouldKill bool) error {
+	c.stopLock.Lock()
+	defer c.stopLock.Unlock()
+	if c.stopping {
+		return errors.New("cannot register an exec PID: container is stopping")
+	}
+	c.execPIDs[pid] = shouldKill
+	return nil
+}
+
+// DeleteExecPID is for deregistering a pid after it has exited.
+func (c *Container) DeleteExecPID(pid int) {
+	c.stopLock.Lock()
+	defer c.stopLock.Unlock()
+	delete(c.execPIDs, pid)
+}
+
+// KillExecPIDs loops through the saved execPIDs and sends a signal to them.
+// If shouldKill is true, the signal is SIGKILL. Otherwise, SIGINT.
+func (c *Container) KillExecPIDs() {
+	c.stopLock.Lock()
+	toKill := c.execPIDs
+	c.stopLock.Unlock()
+
+	for len(toKill) != 0 {
+		unkilled := map[int]bool{}
+		for pid, shouldKill := range toKill {
+			sig := syscall.SIGINT
+			if shouldKill {
+				sig = syscall.SIGKILL
+			}
+			if err := syscall.Kill(pid, sig); err != nil && !errors.Is(err, syscall.ESRCH) {
+				unkilled[pid] = shouldKill
```

Comment:
So, you can either kill a process or not, and it doesn't make sense to retry. From kill(2) man page:
```
ERRORS
       EINVAL An invalid signal was specified.

       EPERM  The calling process does not have permission to send the signal to any of the target processes.

       ESRCH  The target process or process group does not exist.  Note that an existing process might be a zombie, a process that has terminated execution, but has not yet been wait(2)ed for.
```

Also, kill is async, meaning it can take some time for the kernel to actually release the process (and, in case of SIGINT, it also depends on what the app is doing).

There are two ways to make kill synchronous, if you need that here. Alas neither way is perfect.
1. Keep killing pid(s) till you get ESRCH (with some sensible timeout). The downside is there's a chance of PID wrap issue, meaning you'll be killing a wrong process. Practically, if the time the killing code is running is small enough (seconds), this issue is highly unlikely to happen.
2. Use pidfd and wait on events. Unfortunately this requires a new kernel and some amount of non-trivial code.

The current code is **kind of** implementing the first approach, but
- it does not look like it (as you keep sending the same signal);
- the number of iterations (and thus the total time spend killing) need to be limited.

So I suggest to rewrite this as following:
 - kill all execs (no loop needed)
 - check that they are killed (sending signal 0 in a loop with limited number of tries)
 - (optionally) use KILL for those PIDs that you sent INT to before, as a last resort. This can be done somewhere in the middle of the loop.

Reply:
* return (or log) an error if some execs still can't be killed.

Reply:
Something that I ran through @haircommander recently.

We could also leverage setting the `Setpgid` via `syscall.SysProcAttr{}` to first deliver a signal to the given process group. We do this today when creating the main container process but not using an Exec-type approach.

There is also `unix.PR_SET_PDEATHSIG` we could set to use `unix.SIGKILL` using `unix.Prctl()` better clean up children, etc.

If we want to tackle the PID wrap issue, then to avoid having to track everything internally, we could inject an environment variable such as `_OCI_COOKIE` or simply `CRIO` with some value (e.g., a random SHA256 or UUID generated at the CRI-O startup) and then look for this when we are about to kill a given PID.

Reply:
> The downside is there's a chance of PID wrap issue, meaning you'll be killing a wrong process. Practically, if the time the killing code is running is small enough (seconds), this issue is highly unlikely to happen.

yeah I don't think we need to worry about PID wrap in this case